File: compatibility-tests/sdk-test-go/tests/s3_test.go
package tests

import (
	"bytes"
	"context"
	"net/url"
	"strings"
	"testing"

	"floci-sdk-test-go/internal/testutil"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	s3types "github.com/aws/aws-sdk-go-v2/service/s3/types"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestS3(t *testing.T) {
	ctx := context.Background()
	svc := testutil.S3Client()
	bucket := "go-test-bucket"
	key := "test-object.json"
	content := `{"source":"go-test"}`

	// Cleanup at end
	t.Cleanup(func() {
		svc.DeleteObject(ctx, &s3.DeleteObjectInput{Bucket: aws.String(bucket), Key: aws.String(key)})
		svc.DeleteObject(ctx, &s3.DeleteObjectInput{Bucket: aws.String(bucket), Key: aws.String("copy-" + key)})
		svc.DeleteBucket(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucket)})
	})

	t.Run("CreateBucket", func(t *testing.T) {
		_, err := svc.CreateBucket(ctx, &s3.CreateBucketInput{Bucket: aws.String(bucket)})
		require.NoError(t, err)
	})

	t.Run("ListBuckets", func(t *testing.T) {
		r, err := svc.ListBuckets(ctx, &s3.ListBucketsInput{})
		require.NoError(t, err)
		assert.NotEmpty(t, r.Buckets)
	})

	t.Run("PutObject", func(t *testing.T) {
		_, err := svc.PutObject(ctx, &s3.PutObjectInput{
			Bucket:      aws.String(bucket),
			Key:         aws.String(key),
			Body:        strings.NewReader(content),
			ContentType: aws.String("application/json"),
		})
		require.NoError(t, err)
	})

	t.Run("GetObject", func(t *testing.T) {
		out, err := svc.GetObject(ctx, &s3.GetObjectInput{
			Bucket: aws.String(bucket),
			Key:    aws.String(key),
		})
		require.NoError(t, err)
		defer out.Body.Close()

		var buf bytes.Buffer
		buf.ReadFrom(out.Body)
		assert.Equal(t, content, buf.String())
	})

	t.Run("HeadObject", func(t *testing.T) {
		head, err := svc.HeadObject(ctx, &s3.HeadObjectInput{
			Bucket: aws.String(bucket),
			Key:    aws.String(key),
		})
		require.NoError(t, err)
		assert.NotNil(t, head.LastModified)
		assert.Zero(t, head.LastModified.Nanosecond(), "LastModified should have second precision")
	})

	t.Run("ListObjectsV2", func(t *testing.T) {
		r, err := svc.ListObjectsV2(ctx, &s3.ListObjectsV2Input{Bucket: aws.String(bucket)})
		require.NoError(t, err)
		assert.NotEmpty(t, r.Contents)
	})

	t.Run("CopyObject", func(t *testing.T) {
		_, err := svc.CopyObject(ctx, &s3.CopyObjectInput{
			Bucket:     aws.String(bucket),
			CopySource: aws.String(bucket + "/" + key),
			Key:        aws.String("copy-" + key),
		})
		require.NoError(t, err)
	})

	t.Run("DeleteObject", func(t *testing.T) {
		_, err := svc.DeleteObject(ctx, &s3.DeleteObjectInput{
			Bucket: aws.String(bucket),
			Key:    aws.String(key),
		})
		require.NoError(t, err)
	})

	t.Run("DeleteBucket", func(t *testing.T) {
		// First delete the copy
		svc.DeleteObject(ctx, &s3.DeleteObjectInput{
			Bucket: aws.String(bucket),
			Key:    aws.String("copy-" + key),
		})

		_, err := svc.DeleteBucket(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucket)})
		require.NoError(t, err)
	})
}

func TestS3LocationConstraint(t *testing.T) {
	ctx := context.Background()
	svc := testutil.S3Client()
	euBucket := "go-test-bucket-eu"

	t.Cleanup(func() {
		svc.DeleteBucket(ctx, &s3.DeleteBucketInput{Bucket: aws.String(euBucket)})
	})

	t.Run("CreateBucketWithLocationConstraint", func(t *testing.T) {
		_, err := svc.CreateBucket(ctx, &s3.CreateBucketInput{
			Bucket: aws.String(euBucket),
			CreateBucketConfiguration: &s3types.CreateBucketConfiguration{
				LocationConstraint: s3types.BucketLocationConstraintEuCentral1,
			},
		})
		require.NoError(t, err)
	})

	t.Run("GetBucketLocation", func(t *testing.T) {
		loc, err := svc.GetBucketLocation(ctx, &s3.GetBucketLocationInput{Bucket: aws.String(euBucket)})
		require.NoError(t, err)
		assert.Equal(t, "eu-central-1", string(loc.LocationConstraint))
	})
}

// TestS3NonASCIIKey tests CopyObject with non-ASCII (multibyte) keys.
// Regression test for issue #93.
func TestS3NonASCIIKey(t *testing.T) {
	ctx := context.Background()
	svc := testutil.S3Client()
	bucket := "go-test-non-ascii"
	srcKey := "src/テスト画像.png"
	dstKey := "dst/テスト画像.png"

	t.Cleanup(func() {
		svc.DeleteObject(ctx, &s3.DeleteObjectInput{Bucket: aws.String(bucket), Key: aws.String(srcKey)})
		svc.DeleteObject(ctx, &s3.DeleteObjectInput{Bucket: aws.String(bucket), Key: aws.String(dstKey)})
		svc.DeleteBucket(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucket)})
	})

	// Setup
	_, err := svc.CreateBucket(ctx, &s3.CreateBucketInput{Bucket: aws.String(bucket)})
	require.NoError(t, err)

	_, err = svc.PutObject(ctx, &s3.PutObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(srcKey),
		Body:   strings.NewReader("non-ascii content"),
	})
	require.NoError(t, err)

	// Copy with non-ASCII key
	_, err = svc.CopyObject(ctx, &s3.CopyObjectInput{
		Bucket:     aws.String(bucket),
		CopySource: aws.String(bucket + "/" + url.PathEscape(srcKey)),
		Key:        aws.String(dstKey),
	})
	require.NoError(t, err, "CopyObject with non-ASCII key should succeed")

	// Verify the copy exists
	_, err = svc.HeadObject(ctx, &s3.HeadObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(dstKey),
	})
	require.NoError(t, err, "copied object should exist")
}

// TestS3MultipartCopyNonASCIIKey exercises UploadPartCopy with a URL-encoded
// non-ASCII source key to cover the multipart copy code path.
func TestS3MultipartCopyNonASCIIKey(t *testing.T) {
	ctx := context.Background()
	svc := testutil.S3Client()
	bucket := "go-test-multipart-copy-non-ascii"
	srcKey := "src/テスト画像.bin"
	dstKey := "dst/テスト画像.bin"
	content := bytes.Repeat([]byte("a"), 6*1024*1024)
	var uploadID string

	t.Cleanup(func() {
		if uploadID != "" {
			svc.AbortMultipartUpload(ctx, &s3.AbortMultipartUploadInput{
				Bucket:   aws.String(bucket),
				Key:      aws.String(dstKey),
				UploadId: aws.String(uploadID),
			})
		}
		svc.DeleteObject(ctx, &s3.DeleteObjectInput{Bucket: aws.String(bucket), Key: aws.String(srcKey)})
		svc.DeleteObject(ctx, &s3.DeleteObjectInput{Bucket: aws.String(bucket), Key: aws.String(dstKey)})
		svc.DeleteBucket(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucket)})
	})

	_, err := svc.CreateBucket(ctx, &s3.CreateBucketInput{Bucket: aws.String(bucket)})
	require.NoError(t, err)

	_, err = svc.PutObject(ctx, &s3.PutObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(srcKey),
		Body:   bytes.NewReader(content),
	})
	require.NoError(t, err)

	createOut, err := svc.CreateMultipartUpload(ctx, &s3.CreateMultipartUploadInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(dstKey),
	})
	require.NoError(t, err)
	uploadID = aws.ToString(createOut.UploadId)
	require.NotEmpty(t, uploadID)

	partOut, err := svc.UploadPartCopy(ctx, &s3.UploadPartCopyInput{
		Bucket:     aws.String(bucket),
		Key:        aws.String(dstKey),
		UploadId:   aws.String(uploadID),
		PartNumber: aws.Int32(1),
		CopySource: aws.String(bucket + "/" + url.PathEscape(srcKey)),
	})
	require.NoError(t, err, "UploadPartCopy with non-ASCII key should succeed")
	require.NotNil(t, partOut.CopyPartResult)
	require.NotNil(t, partOut.CopyPartResult.ETag)

	_, err = svc.CompleteMultipartUpload(ctx, &s3.CompleteMultipartUploadInput{
		Bucket:   aws.String(bucket),
		Key:      aws.String(dstKey),
		UploadId: aws.String(uploadID),
		MultipartUpload: &s3types.CompletedMultipartUpload{
			Parts: []s3types.CompletedPart{
				{
					ETag:       partOut.CopyPartResult.ETag,
					PartNumber: aws.Int32(1),
				},
			},
		},
	})
	require.NoError(t, err)
	uploadID = ""

	getOut, err := svc.GetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(dstKey),
	})
	require.NoError(t, err)
	defer getOut.Body.Close()

	var copied bytes.Buffer
	_, err = copied.ReadFrom(getOut.Body)
	require.NoError(t, err)
	assert.Equal(t, content, copied.Bytes())
}

// TestS3LargeObject tests uploading a 25 MB object.
// Validates upload size limit handling.
func TestS3LargeObject(t *testing.T) {
	ctx := context.Background()
	svc := testutil.S3Client()
	bucket := "go-test-large-upload"
	key := "large-object-25mb.bin"
	size := int64(25 * 1024 * 1024) // 25 MB

	t.Cleanup(func() {
		svc.DeleteObject(ctx, &s3.DeleteObjectInput{Bucket: aws.String(bucket), Key: aws.String(key)})
		svc.DeleteBucket(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucket)})
	})

	// Setup
	_, err := svc.CreateBucket(ctx, &s3.CreateBucketInput{Bucket: aws.String(bucket)})
	require.NoError(t, err)

	// Create 25 MB payload
	payload := make([]byte, size)

	_, err = svc.PutObject(ctx, &s3.PutObjectInput{
		Bucket:        aws.String(bucket),
		Key:           aws.String(key),
		Body:          bytes.NewReader(payload),
		ContentType:   aws.String("application/octet-stream"),
		ContentLength: aws.Int64(size),
	})
	require.NoError(t, err, "PutObject 25 MB should succeed")

	// Verify content-length via HeadObject
	head, err := svc.HeadObject(ctx, &s3.HeadObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
	})
	require.NoError(t, err)
	assert.Equal(t, size, aws.ToInt64(head.ContentLength), "content-length should be 25 MB")
}

File: compatibility-tests/sdk-test-rust/tests/s3_test.rs
mod common;

use aws_sdk_s3::types::{BucketLocationConstraint, CreateBucketConfiguration, Delete, ObjectIdentifier};

#[tokio::test(flavor = "multi_thread")]
async fn test_s3_create_bucket() {
    let s3 = common::s3_client().await;
    let bucket = "rust-test-create-bucket";

    let _guard = common::CleanupGuard::new({
        let s3 = s3.clone();
        async move {
            let _ = s3.delete_bucket().bucket(bucket).send().await;
        }
    });

    let result = s3.create_bucket().bucket(bucket).send().await;
    assert!(result.is_ok(), "CreateBucket failed: {:?}", result.err());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_s3_create_bucket_with_location_constraint() {
    let s3 = common::s3_client().await;
    let bucket = "rust-test-eu-bucket";

    let _guard = common::CleanupGuard::new({
        let s3 = s3.clone();
        async move {
            let _ = s3.delete_bucket().bucket(bucket).send().await;
        }
    });

    let result = s3
        .create_bucket()
        .bucket(bucket)
        .create_bucket_configuration(
            CreateBucketConfiguration::builder()
                .location_constraint(BucketLocationConstraint::EuCentral1)
                .build(),
        )
        .send()
        .await;
    assert!(result.is_ok(), "CreateBucket with location failed: {:?}", result.err());

    // Verify location
    let loc = s3.get_bucket_location().bucket(bucket).send().await;
    assert!(loc.is_ok());
    assert_eq!(
        loc.unwrap()
            .location_constraint()
            .map(|c| c.as_str())
            .unwrap_or(""),
        "eu-central-1"
    );
}

#[tokio::test(flavor = "multi_thread")]
async fn test_s3_list_buckets() {
    let s3 = common::s3_client().await;
    let bucket = "rust-test-list-buckets";

    let _guard = common::CleanupGuard::new({
        let s3 = s3.clone();
        async move {
            let _ = s3.delete_bucket().bucket(bucket).send().await;
        }
    });

    // Setup
    s3.create_bucket().bucket(bucket).send().await.expect("setup");

    let result = s3.list_buckets().send().await;
    assert!(result.is_ok(), "ListBuckets failed: {:?}", result.err());
    assert!(!result.unwrap().buckets().is_empty());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_s3_put_and_get_object() {
    let s3 = common::s3_client().await;
    let bucket = "rust-test-put-get";
    let key = "test.json";
    let content = r#"{"source":"rust-test"}"#;

    let _guard = common::CleanupGuard::new({
        let s3 = s3.clone();
        async move {
            let _ = s3.delete_object().bucket(bucket).key(key).send().await;
            let _ = s3.delete_bucket().bucket(bucket).send().await;
        }
    });

    // Setup
    s3.create_bucket().bucket(bucket).send().await.expect("setup");

    // Put
    let put_result = s3
        .put_object()
        .bucket(bucket)
        .key(key)
        .body(bytes::Bytes::from(content).into())
        .content_type("application/json")
        .send()
        .await;
    assert!(put_result.is_ok(), "PutObject failed: {:?}", put_result.err());

    // Get
    let get_result = s3.get_object().bucket(bucket).key(key).send().await;
    assert!(get_result.is_ok(), "GetObject failed: {:?}", get_result.err());

    let body = get_result
        .unwrap()
        .body
        .collect()
        .await
        .map(|b| b.into_bytes());
    assert!(body.is_ok());
    assert_eq!(body.unwrap().as_ref(), content.as_bytes());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_s3_head_object() {
    let s3 = common::s3_client().await;
    let bucket = "rust-test-head";
    let key = "head-test.txt";

    let _guard = common::CleanupGuard::new({
        let s3 = s3.clone();
        async move {
            let _ = s3.delete_object().bucket(bucket).key(key).send().await;
            let _ = s3.delete_bucket().bucket(bucket).send().await;
        }
    });

    // Setup
    s3.create_bucket().bucket(bucket).send().await.expect("setup");
    s3.put_object()
        .bucket(bucket)
        .key(key)
        .body(bytes::Bytes::from("test").into())
        .send()
        .await
        .expect("put");

    let result = s3.head_object().bucket(bucket).key(key).send().await;
    assert!(result.is_ok(), "HeadObject failed: {:?}", result.err());

    let head = result.unwrap();
    assert!(head.last_modified().is_some());
    // Check second precision
    assert_eq!(head.last_modified().unwrap().subsec_nanos(), 0);
}

#[tokio::test(flavor = "multi_thread")]
async fn test_s3_list_objects_v2() {
    let s3 = common::s3_client().await;
    let bucket = "rust-test-list-objects";
    let key = "list-test.txt";

    let _guard = common::CleanupGuard::new({
        let s3 = s3.clone();
        async move {
            let _ = s3.delete_object().bucket(bucket).key(key).send().await;
            let _ = s3.delete_bucket().bucket(bucket).send().await;
        }
    });

    // Setup
    s3.create_bucket().bucket(bucket).send().await.expect("setup");
    s3.put_object()
        .bucket(bucket)
        .key(key)
        .body(bytes::Bytes::from("test").into())
        .send()
        .await
        .expect("put");

    let result = s3.list_objects_v2().bucket(bucket).send().await;
    assert!(result.is_ok(), "ListObjectsV2 failed: {:?}", result.err());
    assert!(!result.unwrap().contents().is_empty());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_s3_copy_object() {
    let s3 = common::s3_client().await;
    let bucket = "rust-test-copy";
    let src_key = "source.txt";
    let dst_key = "copy.txt";

    let _guard = common::CleanupGuard::new({
        let s3 = s3.clone();
        async move {
            let _ = s3
                .delete_objects()
                .bucket(bucket)
                .delete(
                    Delete::builder()
                        .objects(ObjectIdentifier::builder().key(src_key).build().unwrap())
                        .objects(ObjectIdentifier::builder().key(dst_key).build().unwrap())
                        .build()
                        .unwrap(),
                )
                .send()
                .await;
            let _ = s3.delete_bucket().bucket(bucket).send().await;
        }
    });

    // Setup
    s3.create_bucket().bucket(bucket).send().await.expect("setup");
    s3.put_object()
        .bucket(bucket)
        .key(src_key)
        .body(bytes::Bytes::from("source").into())
        .send()
        .await
        .expect("put");

    let copy_source = format!("{}/{}", bucket, src_key);
    let result = s3
        .copy_object()
        .bucket(bucket)
        .copy_source(&copy_source)
        .key(dst_key)
        .send()
        .await;
    assert!(result.is_ok(), "CopyObject failed: {:?}", result.err());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_s3_delete_bucket() {
    let s3 = common::s3_client().await;
    let bucket = "rust-test-delete-bucket";

    // No cleanup guard needed - test is about deletion

    // Setup
    s3.create_bucket().bucket(bucket).send().await.expect("setup");

    let result = s3.delete_bucket().bucket(bucket).send().await;
    assert!(result.is_ok(), "DeleteBucket failed: {:?}", result.err());
}

/// Percent-encodes non-ASCII bytes in an S3 key, preserving / as a path separator.
/// The Rust SDK does not URL-encode CopySource headers, so this must be done manually.
fn s3_encode_key(key: &str) -> String {
    let mut out = String::with_capacity(key.len());
    for b in key.bytes() {
        match b {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' | b'/' => {
                out.push(b as char);
            }
            _ => {
                out.push_str(&format!("%{:02X}", b));
            }
        }
    }
    out
}

/// Regression test for issue #93: CopyObject with non-ASCII (multibyte) key.
/// The Rust SDK does not URL-encode CopySource headers; encode the key manually.
#[tokio::test(flavor = "multi_thread")]
async fn test_s3_copy_object_non_ascii_key() {
    let s3 = common::s3_client().await;
    let bucket = "rust-test-non-ascii";
    let src_key = "src/テスト画像.png";
    let dst_key = "dst/テスト画像.png";

    let _guard = common::CleanupGuard::new({
        let s3 = s3.clone();
        async move {
            let _ = s3
                .delete_objects()
                .bucket(bucket)
                .delete(
                    Delete::builder()
                        .objects(ObjectIdentifier::builder().key(src_key).build().unwrap())
                        .objects(ObjectIdentifier::builder().key(dst_key).build().unwrap())
                        .build()
                        .unwrap(),
                )
                .send()
                .await;
            let _ = s3.delete_bucket().bucket(bucket).send().await;
        }
    });

    // Setup
    s3.create_bucket().bucket(bucket).send().await.expect("create bucket");
    s3.put_object()
        .bucket(bucket)
        .key(src_key)
        .body(bytes::Bytes::from("non-ascii content").into())
        .send()
        .await
        .expect("put source object");

    // Copy with URL-encoded non-ASCII key
    let copy_source = format!("{}/{}", bucket, s3_encode_key(src_key));
    let result = s3
        .copy_object()
        .bucket(bucket)
        .copy_source(&copy_source)
        .key(dst_key)
        .send()
        .await;

    assert!(
        result.is_ok(),
        "CopyObject with non-ASCII key failed: {:?}",
        result.err()
    );

    // Verify the copy exists
    let head = s3.head_object().bucket(bucket).key(dst_key).send().await;
    assert!(head.is_ok(), "copied object should exist");
}

/// Test large object upload (25 MB) - validates upload size limit handling.
#[tokio::test(flavor = "multi_thread")]
async fn test_s3_put_object_25mb() {
    let s3 = common::s3_client().await;
    let bucket = "rust-test-large-upload";
    let key = "large-object-25mb.bin";
    let size: i64 = 25 * 1024 * 1024; // 25 MB

    let _guard = common::CleanupGuard::new({
        let s3 = s3.clone();
        async move {
            let _ = s3.delete_object().bucket(bucket).key(key).send().await;
            let _ = s3.delete_bucket().bucket(bucket).send().await;
        }
    });

    // Setup
    s3.create_bucket().bucket(bucket).send().await.expect("create bucket");

    // Create 25 MB payload
    let payload = vec![0u8; size as usize];

    // Upload
    let put_result = s3
        .put_object()
        .bucket(bucket)
        .key(key)
        .body(bytes::Bytes::from(payload).into())
        .content_type("application/octet-stream")
        .content_length(size)
        .send()
        .await;

    assert!(
        put_result.is_ok(),
        "PutObject 25 MB failed: {:?}",
        put_result.err()
    );

    // Verify content-length via HeadObject
    let head_result = s3.head_object().bucket(bucket).key(key).send().await;
    assert!(
        head_result.is_ok(),
        "HeadObject failed: {:?}",
        head_result.err()
    );

    let head = head_result.unwrap();
    assert_eq!(
        head.content_length(),
        Some(size),
        "content-length should be {} bytes",
        size
    );
}
